kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/ModalContainer.kt

```diff
@@ -242,3 +255,28 @@ abstract class ModalContainer<ModalRenderingT : Any>
     )
   }
 }
+
+private class OnDestroy(private val block: () -> Unit) : LifecycleObserver {
+  @OnLifecycleEvent(ON_DESTROY)
+  fun onDestroy() = block()
+}
+
+private val Dialog.activity: AppCompatActivity
+  get() = decorView.context.activity
+
+private val Dialog.decorView: View
+  get() = window!!.decorView
+
+private val Context.activity: AppCompatActivity
+  get() {
+    var wrapper = this as ContextWrapper
+    var activity: Activity? = null
+    do {
+      if (wrapper is Activity) {
+        activity = wrapper
+      } else {
+        wrapper = wrapper.baseContext as ContextWrapper
+      }
+    } while (activity == null)
+    return activity as AppCompatActivity
```

Comment:
Kind of late for this cast, should be using it in the loop.

Reply:
Could write this more concisely as a tailrec function:
```kotlin
tailrec fun Context.activity(): AppCompatActivity {
  if (this is Activity) return this
  return (this as ContextWrapper).baseContext.activity()
}
```

Reply:
```kotlin
tailrec fun Context.activity(): AppCompatActivity? = when {
  this is Activity -> this as? AppCompatActivity
  else -> (this as? ContextWrapper)?.baseContext?.activity()
}
```